File: apps/web/src/lib/points.test.ts
import { describe, expect, it } from "vitest";
import { pointsEarned, pointsToSpend, pointsValueCents } from "./points";

describe("pointsEarned", () => {
  it("earns on the sale total at the shop rate — Rs 1,150 at 1 per Rs 100 is 11", () => {
    expect(pointsEarned({ totalCents: 115_000, pointsPaidCents: 0, pointsPer100: 1 })).toBe(11);
  });

  it("ignores the share settled with points — the exclusion bites at the next rupee, not this one", () => {
    // Same Rs 1,150 bill, Rs 50 of it already paid in points. Rs 1,100 earning at
    // 1 per Rs 100 is still 11 — the flooring doesn't move until Rs 1,200.
    expect(pointsEarned({ totalCents: 115_000, pointsPaidCents: 5_000, pointsPer100: 1 })).toBe(11);
  });

  it("rounds down — a part point is not a point", () => {
    expect(pointsEarned({ totalCents: 9_900, pointsPaidCents: 0, pointsPer100: 1 })).toBe(0);
  });

  it("honours a rate above one", () => {
    expect(pointsEarned({ totalCents: 100_000, pointsPaidCents: 0, pointsPer100: 2.5 })).toBe(25);
  });

  it("earns nothing on a bill settled entirely in points", () => {
    expect(pointsEarned({ totalCents: 50_000, pointsPaidCents: 50_000, pointsPer100: 1 })).toBe(0);
  });

  it("a zero rate earns nothing, without throwing", () => {
    expect(pointsEarned({ totalCents: 100_000, pointsPaidCents: 0, pointsPer100: 0 })).toBe(0);
  });

  it("a negative rate earns nothing, without throwing", () => {
    expect(pointsEarned({ totalCents: 100_000, pointsPaidCents: 0, pointsPer100: -1 })).toBe(0);
  });
});

describe("pointsToSpend", () => {
  it("rounds up — the shop is never out of pocket for a fraction of a point", () => {
    expect(pointsToSpend(5_050, 1)).toBe(51);
  });

  it("honours a point worth more than a rupee", () => {
    expect(pointsToSpend(10_000, 5)).toBe(20);
  });
});

describe("pointsValueCents", () => {
  it("states what a balance is worth", () => {
    expect(pointsValueCents(120, 1)).toBe(12_000);
  });
});

File: android/app/src/test/java/mu/carfection/pos/core/money/PointsTest.kt
package mu.carfection.pos.core.money

import org.junit.Assert.assertEquals
import org.junit.Test

/**
 * Same cases as apps/web/src/lib/points.test.ts, same numbers — the parity gate diffs the
 * two suites, and a case present on one side only is drift.
 */
class PointsTest {

    // ── pointsEarned ─────────────────────────────────────────────────────────

    @Test
    fun `earns on the sale total at the shop rate — Rs 1,150 at 1 per Rs 100 is 11`() {
        assertEquals(11, pointsEarned(EarnInput(totalCents = 115_000L, pointsPaidCents = 0L, pointsPer100 = 1.0)))
    }

    @Test
    fun `ignores the share settled with points — the exclusion bites at the next rupee, not this one`() {
        // Same Rs 1,150 bill, Rs 50 of it already paid in points. Rs 1,100 earning at
        // 1 per Rs 100 is still 11 — the flooring doesn't move until Rs 1,200.
        assertEquals(11, pointsEarned(EarnInput(totalCents = 115_000L, pointsPaidCents = 5_000L, pointsPer100 = 1.0)))
    }

    @Test
    fun `rounds down — a part point is not a point`() {
        assertEquals(0, pointsEarned(EarnInput(totalCents = 9_900L, pointsPaidCents = 0L, pointsPer100 = 1.0)))
    }

    @Test
    fun `honours a rate above one`() {
        assertEquals(25, pointsEarned(EarnInput(totalCents = 100_000L, pointsPaidCents = 0L, pointsPer100 = 2.5)))
    }

    @Test
    fun `earns nothing on a bill settled entirely in points`() {
        assertEquals(0, pointsEarned(EarnInput(totalCents = 50_000L, pointsPaidCents = 50_000L, pointsPer100 = 1.0)))
    }

    @Test
    fun `a zero rate earns nothing, without throwing`() {
        assertEquals(0, pointsEarned(EarnInput(totalCents = 100_000L, pointsPaidCents = 0L, pointsPer100 = 0.0)))
    }

    @Test
    fun `a negative rate earns nothing, without throwing`() {
        assertEquals(0, pointsEarned(EarnInput(totalCents = 100_000L, pointsPaidCents = 0L, pointsPer100 = -1.0)))
    }

    // ── pointsToSpend ────────────────────────────────────────────────────────

    @Test
    fun `rounds up — the shop is never out of pocket for a fraction of a point`() {
        assertEquals(51, pointsToSpend(5_050L, 1.0))
    }

    @Test
    fun `honours a point worth more than a rupee`() {
        assertEquals(20, pointsToSpend(10_000L, 5.0))
    }

    // ── pointsValueCents ─────────────────────────────────────────────────────

    @Test
    fun `states what a balance is worth`() {
        assertEquals(12_000L, pointsValueCents(120, 1.0))
    }
}
